### PR TITLE
HCAP- 412 Hired participant withdrawal

### PR DESCRIPTION
**Pull request by dbayly-freshworks**
Changes 
* Added a new column on the hired participants tab. It has an 'Archive' button that opens up a modal. 
* Added Modal to handle form logic for Archive participant flow
* Added validation for both front and back end 
* Archived participants are also visible in the "Archived Participants" tab 
Screen Shot 2021-06-16 at 3 26 56 PM
Screen Shot 2021-06-16 at 3 26 49 PM
Screen Shot 2021-06-16 at 3 26 41 PM

### CODE REVIEW

File: cypress/integration/participantTable.spec.js [hunk 1]
```diff
@@ -15,9 +15,11 @@ describe('Participant Table', () => {
       cy.kcNavAs(role.name, 'participant-view');
       Object.entries(role.tableTabs).forEach((tab) => {
         const [tabText, tabHeaders] = [...tab];
+
         const hasActions = role.tabsWithActions.includes(tabText);
         const columnCount = hasActions ? tabHeaders.length + 1 : tabHeaders.length;
-
+        console.log(tabHeaders);
```
Comment (arranfw) [inline cypress/integration/participantTable.spec.js:21]: Stray console.logs

File: client/src/pages/private/ParticipantTable.js [hunk 2]
```diff
@@ -400,43 +401,23 @@ export default () => {
...
-          !oldColumns.find((column) => column.id === 'status')
-        ) {
-          // Remove statusInfo colum
-          oldColumns = oldColumns.filter((column) => column.id !== 'statusInfo');
-          return [
-            ...oldColumns.slice(0, 3),
-            { id: 'status', name: 'Status' },
-            ...oldColumns.slice(3),
-          ];
+        switch (reducerState.tabValue) {
```
Comment (arranfw) [inline client/src/pages/private/ParticipantTable.js:404]: This looks like it's going to conflict pretty heavily with HCAP-660, we should discuss

File: client/src/constants/participantTableColumnConstants.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
+export const AvailableParticipantColumns = [
```
Comment (arranfw) [inline client/src/constants/participantTableColumnConstants.js:1]: We'll have to see how this relates to HCAP-660 as well

**Review by arranfw — CHANGES_REQUESTED** [review]

**Review by fwpushan — CHANGES_REQUESTED** [review]
Please check my comments. Functional portion looks good to me. Also, for Hire Candidate with Withdrawn status [HCAP-354](https://freshworks.atlassian.net/browse/HCAP-354), the archive option should be disabled.

File: client/src/components/modal-forms/ArchiveHiredParticipantForm.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import { Button } from '../generic';
+import { Field, Formik, Form as FormikForm } from 'formik';
+import { RenderSelectField, RenderDateField, RenderCheckbox } from '../fields';
+import { archiveReasonOptopns, archiveStatusOptions } from '../../constants';
+
+const statusOptions = archiveStatusOptions.map((option) => ({ value: option, label: option }));
+const reasonOptions = archiveReasonOptopns.map((option) => ({ value: option, label: option }));
+const typeOptions = [
+  { value: '', label: '' },
+  { value: 'employmentEnded', label: 'Employment ended' },
```
Comment (fwpushan) [inline client/src/components/modal-forms/ArchiveHiredParticipantForm.js:13]: For consistency purpose, I suggest move this options to constant file

File: server/server.js [hunk 5]
```diff
@@ -327,6 +327,9 @@ app.post(
   keycloak.getUserInfoMiddleware(),
   asyncMiddleware(async (req, res) => {
     await validate(ParticipantStatusChange, req.body);
+    if (req.body.status === 'archived') {
```
Comment (fwpushan) [inline server/server.js:330]: I always suggest immutability of req object
`const status = req.body.status === 'archive' ? 'rejected' : req.body.status`

File: client/src/constants/participantTableColumnConstants.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  { id: 'phoneNumber', name: 'Phone Number' },
+  { id: 'emailAddress', name: 'Email Address' },
+  { id: 'preferredLocation', name: 'Preferred Region(s)' },
+  { id: 'distance', name: 'Site Distance' },
+  { id: 'nonHCAP', name: 'Non-HCAP' },
+  { id: 'userUpdatedAt', name: 'Last Updated' },
+  { id: 'engage' },
+];
+
+export const HiredCandidateColumns = [
```
[the remark this answers is not in the document]
Comment (fwpushan) [reply client/src/constants/participantTableColumnConstants.js:46]: Please discard above comment. 
1. `already_hired` is not applicable for Hired column 
2. `Withdrawn` is handled by your changes

File: server/server.js (hunk 5 above)
Comment (dbayly-freshworks) [reply server/server.js]: I ended up removing this part and just adding a new status

File: client/src/constants/participantTableColumnConstants.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  { id: 'emailAddress', name: 'Email Address' },
+  { id: 'preferredLocation', name: 'Preferred Region(s)' },
+  { id: 'distance', name: 'Site Distance' },
+  { id: 'nonHCAP', name: 'Non-HCAP' },
+  { id: 'callbackStatus', name: 'Callback Status' },
+  { id: 'userUpdatedAt', name: 'Last Updated' },
+  { id: 'engage' },
+];
+
+export const MyCandidatesColumns = [
```
[the remark this answers is not in the document]
Comment (fwpushan) [reply client/src/constants/participantTableColumnConstants.js:16]: Please ignore this comment as per our discussion.

**Review by fwpushan — COMMENTED** [review]
LGTM, please check formatting.

**Review by arranfw — APPROVED** [review]

**Review by fwpushan — APPROVED** [review]

**Pull request merged**